core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -152,8 +152,8 @@ private object RTS {
     final def apply(v: Any): IO[Any, Any] = IO.now(v)
   }
 
-  final case class Finalizer[E](finalizer: ExitResult[E, Any] => IO[Void, Unit]) extends Function[Any, IO[E, Any]] {
-    final def apply(v: Any): IO[E, Any] = IO.now(v)
+  final case class Finalizer[E](finalizer: Infallible[Unit]) extends Function[Any, IO[E, Any]] {
+    final def apply(v: Any): IO[E, Any] = finalizer.widenError[E] *> IO.now(v)
```

Comment:
This is the tricky case where, in the "happy" path (no errors in the body), interruption of the finalizer is possible (but should not be possible).

One way is to solve this in `nextInstr`—if we see a `Finalizer` coming off the stack, we `++` the `noInterrupt` and push a `_ => exitUninterruptible` onto the stack. It will be slower though. In any case, maybe best dealt with in a separate PR. It's definitely a problem though.

Reply:
I see. Couldn't this be done in `Finalizer.apply` itself? It's only ever called in `nextInstr` afaics. Then we could keep the `Ensuring` branch as is and avoid a pattern match.
```scala
final case class Finalizer[E](finalizer: Infallible[Unit]) extends Function[Any, IO[E, Any]] {
  @inline def noInterrupt = enterUninterruptible.flatMap(_ => finalizer *> exitUninterruptible)
  final def apply(v: Any): IO[E, Any] = noInterrupt *> IO.now(v)
}
```

Reply:
No, because `Finalizer` will be popped off the stack as `curIo`, then the next time through the `while` loop, interruption will be checked. At this point, if it's interrupted, the whole `Finalizer` will be off the stack and onto `curIo`, so when `interruptStack` is called, it won't be cleaned up properly.